src/cpp/rtps/writer/StatefulWriter.cpp

```diff
@@ -437,12 +437,11 @@ bool StatefulWriter::intraprocess_heartbeat(
         bool liveliness)
 {
     bool returned_value = false;
+    LocalReaderPointer::Instance local_reader = reader_proxy->local_reader();
 
-    std::lock_guard<RecursiveTimedMutex> guardW(mp_mutex);
-    RTPSReader* reader = RTPSDomainImpl::find_local_reader(reader_proxy->guid());
-
-    if (reader)
+    if (local_reader)
     {
+        std::lock_guard<RecursiveTimedMutex> guardW(mp_mutex);
```

Comment:
Even though this guard is not new, so no deadlock would be introduced by the changes here, I would unlock it before calling `process_heartbeat_msg` on the reader. This would require copying `heartbeat_count_` into a local variable before unlocking the mutex.